### PR TITLE
Add force_reconnect_interval config to Wazuh Agent

### PR DESCRIPTION
**Pull request by palaciosjeremias**
|Related issue|
|---|
|#7893|

## Description
This PR adds **force_reconnect_interval** config to the agent. 
This new configuration is **disabled** by default and isn't included in **ossec.conf** out of the box. Setting it to **0** or excluding it from **ossec.conf** means **disabled**.
This feature accepts (w|d|h|m|s) format. Setting it as <force_reconnect_interval>2m</force_reconnect_interval> means that the agent will close the connection with the manager and reconnect every 120 seconds. This feature has a minimum accepted value of 1 second.

When this feature is enabled, one INFO log will be written to inform the user that the feature is enabled and the reconnection period. 

This reconnection is explicitly written in the logs with **INFO** tag. This is done this way because the disconnection and connection to a manager will always log an INFO message informing this action, so, it will be clear for users to know the reason for this.

Also, this reconnection sets a lock and a **NACTIVE** status on the agent. This is required because we need to block any thread capable to send data to the manager to avoid it tries to send it during the reconnection. Leading to errors and WARNING messages like `WARNING: (1218): Unable to send message to 'server': No such file or directory`
Setting this lock, have a small possibility of unchaining the following WARNING message if a thread was already going to send data when the reconnection occurs: `WARNING: Process locked due to agent is offline. Waiting for connection...`

## Logs/Alerts example
image

image

## Tests

- Compilation without warnings in every supported platform
  - [X] Linux
  - [X] Windows
- [X] Source installation
- [X] Package installation

- Memory tests for Linux
  - [X] Scan-build report
  - [X] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

**Comment by palaciosjeremias** [comment]
**scan-build report:**
image
**valgrind report:**
image

File: src/headers/string_op.h [hunk 1]
```diff
@@ -160,6 +161,35 @@ int w_parse_bool(const char * string);
  */
 long w_parse_time(const char * string);
 
+/**
+ * @brief Convert seconds into the greater valid time unit (s|m|h|d|w).
+ * The convertion will always round down the output.
```
Comment (DProvinciani) [inline src/headers/string_op.h:166]: Just a typo... `conversion`.

File: src/headers/string_op.h [hunk 2]
```diff
@@ -160,6 +161,35 @@ int w_parse_bool(const char * string);
...
+ *                    TRUE: long format ("second(s)").
+ *                    FALSE: short format ("s")
+ * @return String with the time unit.
+ * @retval "invalid" if the input is negative. A time unit if the input is valid.
+ */
+char* w_seconds_to_time_unit(long seconds, bool long_format);
+
+/**
+ * @brief Convert seconds into the greater time value.
+ *  * The convertion will always round down the output.
```
Comment (DProvinciani) [inline src/headers/string_op.h:185]: Ditto

File: src/shared/string_op.c [hunk 3]
```diff
@@ -759,6 +760,54 @@ long w_parse_time(const char * string) {
     return seconds >= 0 ? seconds : -1;
 }
 
+// Get time unit from seconds
+
+char*  w_seconds_to_time_unit(long seconds, bool long_format) {
+
+    if (seconds < 0) {
+        return "invalid";
```
Comment (DProvinciani) [inline src/shared/string_op.c:768]: Shouldn't be useful to have all these strings defined in the header file? I think that probably could be used by the logs in some cases.

File: src/shared/string_op.c [hunk 4]
```diff
@@ -759,6 +760,54 @@ long w_parse_time(const char * string) {
...
+}
+
+// Get time value from seconds
+
+long w_seconds_to_time_value(long seconds) {
+    
+    if(seconds < 0) {
+        return -1;
+    }
+    else if (seconds >= 604800) {
```
Comment (DProvinciani) [inline src/shared/string_op.c:794]: I suggest also define these constants in a header file.

File: src/win32/win_utils.c [hunk 5]
```diff
@@ -105,6 +105,10 @@ int local_start()
         minfo("Max time to reconnect can't be less than notify_time(%d), using notify_time*3 (%d)", agt->notify_time, agt->max_time_reconnect_try);
     }
     minfo("Using notify time: %d and max time to reconnect: %d", agt->notify_time, agt->max_time_reconnect_try);
+    if (agt->force_reconnect_interval) {
+        minfo("Using force_reconnect_interval, Wazuh Agent will reconnect every %ld %s", \
```
Comment (DProvinciani) [inline src/win32/win_utils.c:109]: The same suggestion of log criteria unification.

**Review by DProvinciani — CHANGES_REQUESTED** [review]
Please, apply changes according to the comments.

File: src/client-agent/agentd.c [hunk 6]
```diff
@@ -70,6 +70,9 @@ void AgentdStart(int uid, int gid, const char *user, const char *group)
     }
 
     minfo("Using notify time: %d and max time to reconnect: %d", agt->notify_time, agt->max_time_reconnect_try);
+    if (agt->force_reconnect_interval) {
+        minfo("Using force_reconnect_interval, Wazuh Agent will reconnect every %ld %s", w_seconds_to_time_value(agt->force_reconnect_interval), w_seconds_to_time_unit(agt->force_reconnect_interval, TRUE));
```
[the remark this answers is not in the document]
Comment (palaciosjeremias) [reply src/client-agent/agentd.c:74]: I agree that using the explicit setting name is more clear. But I think we shouldn't modify this type of legacy logs as the users sometimes use them. Thus, I will use the same format of line 72 in line 74.

**Review by pereyra-m — APPROVED** [review]
Great !

File: src/headers/string_op.h (hunk 1 above)
Comment (palaciosjeremias) [reply src/headers/string_op.h:166]: Fixed

File: src/headers/string_op.h (hunk 2 above)
Comment (palaciosjeremias) [reply src/headers/string_op.h:185]: Fixed

File: src/shared/string_op.c (hunk 3 above)
Comment (palaciosjeremias) [reply src/shared/string_op.c:768]: Done. Every string and numeric value are now defined in **string_op.h** as a conversion utility

File: src/shared/string_op.c (hunk 4 above)
Comment (palaciosjeremias) [reply src/shared/string_op.c:794]: Done

File: src/win32/win_utils.c (hunk 5 above)
Comment (palaciosjeremias) [reply src/win32/win_utils.c:109]: Done

**Review by DProvinciani — APPROVED** [review]
Looks good!

**Pull request merged**